lms/djangoapps/instructor/services.py

```diff
@@ -78,13 +76,60 @@ def delete_student_attempt(self, student_identifier, course_id, content_id, requ
             except (StudentModule.DoesNotExist, enrollment.sub_api.SubmissionError):
                 err_msg = (
                     'Error occurred while attempting to reset student attempts for user '
-                    '{student_identifier} for content_id {content_id}.'.format(
-                        student_identifier=student_identifier,
-                        content_id=content_id
-                    )
+                    f'{student_identifier} for content_id {content_id}.'
                 )
                 log.error(err_msg)
 
+    def complete_student_attempt(self, user_identifier, content_id):
+        """
+        """
+        try:
+            user = get_user_by_username_or_email(user_identifier)
+        except ObjectDoesNotExist:
+            err_msg = (
+                'Error occurred while attempting to complete student attempt for user '
+                f'{user_identifier} for content_id {content_id}. '
+                'User does not exist!'
+            )
+            log.error(err_msg)
+            return
+
+        try:
+            block_key = UsageKey.from_string(content_id)
+        except InvalidKeyError:
+            err_msg = (
+                'Error occurred while attempting to complete student attempts for user '
+                f'{user_identifier} for content_id {content_id}. '
+                'Invalid content_id!'
+            )
+            log.error(err_msg)
+            return
+
+        def _submit_completions(block, user):
+            """
+            """
+            mode = XBlockCompletionMode.get_mode(block)
+            if mode == XBlockCompletionMode.COMPLETABLE:
+                block.runtime.publish(block, 'completion', {'completion': 1.0, 'user_id': user.id})
+            elif mode == XBlockCompletionMode.AGGREGATOR:
+                block_children = ((hasattr(block, 'get_child_descriptors') and block.get_child_descriptors())
+                                 or (hasattr(block, 'get_children') and block.get_children()))
```

Comment:
Is this because of content libraries? Maybe add a comment about what's happening here (you want to get all the content library children, not just the ones for the user -- if that's what's actually happening)

Reply:
It's definitely content libraries and maybe other special blocks as well. But it's actually your assumption flipped. This is so we get only the ones for the user rather than all of the possible children. I figured it was cleaner to only get the children you see rather than all of them and not just mark them all as complete in the completion service. I'll add a comment though about why this is necessary.

As for the whole child_descriptors vs children, I remember being bit by this before (I borrowed this logic from completion where I made a similar go to children function). So I'm not sure when/why one of them is defined  or preferred over the other, but I know just using one causes issues. And then the rest is just being cautious.